File: src/main/java/com/github/stars/controller/StatsController.java
package com.github.stars.controller;

import com.github.stars.dto.LanguageStatsDTO;
import com.github.stars.dto.OverviewStatsDTO;
import com.github.stars.dto.OwnerStatsDTO;
import com.github.stars.dto.TimelineStatsDTO;
import com.github.stars.entity.GithubRepo;
import com.github.stars.service.StatsService;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
public class StatsController {

    private final StatsService statsService;

    public StatsController(StatsService statsService) {
        this.statsService = statsService;
    }

    /**
     * 语言统计数据
     */
    @GetMapping("/api/stats/languages")
    public List<LanguageStatsDTO> languageStats() {
        return statsService.getLanguageStats();
    }

    /**
     * 作者统计数据
     */
    @GetMapping("/api/stats/owners")
    public List<OwnerStatsDTO> ownerStats(@RequestParam(defaultValue = "15") int topN) {
        return statsService.getOwnerStats(topN);
    }

    /**
     * 时间线统计数据
     */
    @GetMapping("/api/stats/timeline")
    public List<TimelineStatsDTO> timelineStats() {
        return statsService.getTimelineStats();
    }

    /**
     * 总体概览数据
     */
    @GetMapping("/api/stats/overview")
    public OverviewStatsDTO overviewStats() {
        return statsService.getOverviewStats();
    }

    /**
     * 最受欢迎仓库 Top N
     */
    @GetMapping("/api/stats/top-starred")
    public List<GithubRepo> topStarredRepos(@RequestParam(defaultValue = "10") int topN) {
        return statsService.getTopStarredRepos(topN);
    }

    /**
     * 最近活跃仓库 Top N
     */
    @GetMapping("/api/stats/recent-active")
    public List<GithubRepo> recentActiveRepos(@RequestParam(defaultValue = "10") int topN) {
        return statsService.getRecentActiveRepos(topN);
    }
}

File: backend-nestjs/src/stats/stats.controller.ts
import { Controller, Get, Logger, Query } from '@nestjs/common';
import { StatsService } from './stats.service';

@Controller('api/stats')
export class StatsController {
    private readonly logger = new Logger(StatsController.name);

    constructor(private readonly service: StatsService) {}

    /**
     * 获取编程语言统计
     *
     * @returns 各语言仓库数量及占比
     */
    @Get('languages')
    async languages() {
        return this.service.getLanguageStats();
    }

    /**
     * 获取仓库所有者排名统计
     *
     * @param topN 返回的排名数量，默认15
     * @returns 所有者排名列表
     */
    @Get('owners')
    async owners(@Query('topN') topN: string) {
        return this.service.getOwnerStats(parseInt(topN) || 15);
    }

    /**
     * 获取 Star 时间线统计
     *
     * @returns 按月份聚合的 Star 数量增长趋势
     */
    @Get('timeline')
    async timeline() {
        return this.service.getTimelineStats();
    }

    /**
     * 获取整体概览统计
     *
     * @returns 仓库总数、Star/Fork 总数、语言/所有者种类数
     */
    @Get('overview')
    async overview() {
        return this.service.getOverviewStats();
    }

    /**
     * 获取 Star 数量排行榜
     *
     * @param topN 返回的仓库数量，默认10
     * @returns 按 starsCount 降序排列的仓库列表
     */
    @Get('top-starred')
    async topStarred(@Query('topN') topN: string) {
        return this.service.getTopStarred(parseInt(topN) || 10);
    }

    /**
     * 获取最近活跃仓库列表
     *
     * @param topN 返回的仓库数量，默认10
     * @returns 按 repoUpdatedAt 降序排列的仓库列表
     */
    @Get('recent-active')
    async recentActive(@Query('topN') topN: string) {
        return this.service.getRecentActive(parseInt(topN) || 10);
    }
}
